### PR TITLE
added solver type field

### PR DESCRIPTION
**Pull request by drozdvadym**
According  to #98 I tried to add solver_type parameter. 
Please check.

### CODE REVIEW

File: digits/model/tasks/caffe_train.py [hunk 1]
```diff
@@ -279,6 +279,7 @@ def save_prototxt_files(self):
         ### Write solver file
 
         solver = caffe_pb2.SolverParameter()
+        solver.solver_type = getattr(solver, self.solver_type)
```
Comment (lukeyeager) [inline digits/model/tasks/caffe_train.py]: Does this actually work? Don't you want this instead:

```python
solver.solver_type = self.solver_type
```

File: digits/templates/models/images/classification/new.html [hunk 2]
```diff
@@ -106,6 +106,18 @@
...
+                    <div class="input-group">
+                        {{form.solver_type(class='form-control')}}
+                        <span name="solver_type_explanation"
+                            class="input-group-addon explanation-tooltip glyphicon glyphicon-question-sign"
+                            data-container="body"
+                            title="What type of solver will be used??"
+                            ></span>
+                    </div>
+                </div>
+
```
Comment (lukeyeager) [inline digits/templates/models/images/classification/new.html]: Can we move this below "Batch size" instead?

File: digits/model/forms.py [hunk 3]
```diff
@@ -81,6 +81,17 @@ def validate_NetParameter(form, field):
...
+    ### Solver types
+
+    solver_type = wtforms.SelectField('Solver type',
+        choices = [
+                ('SGD', 'Stochastic gradient descent'),
+                ('ADAGRAD', 'Adaptive gradient'),
+                ('NESTEROV', 'Nesterov`s accelerated gradient'),
+                ],
+            default = 'SGD'
+            )
```
Comment (lukeyeager) [inline digits/model/forms.py]: How about this for the names:

    ('SGD', 'Stochastic gradient descent (SGD)'),
    ('ADAGRAD', 'Adaptive gradient (AdaGrad)'),
    ('NESTEROV', "Nesterov's accelerated gradient"),

**Comment by lukeyeager** [comment]
Thanks for contributing! Major props for fixing somebody else's issue. I see one issue with this besides the comments above:

When I choose AdaGrad, caffe gives me this error:

    0 == this->param_.momentum() (0 vs. 0.9) Momentum cannot be used with AdaGrad.

I think you'll want to change [this line](https://github.com/NVIDIA/DIGITS/blob/70df02f3271d95dbf1aa90be6eecb55220a30fff/digits/model/tasks/caffe_train.py#L345) in `caffe_train.py` from
```python
solver.momentum = 0.9
```
to
```python
if solver.solver_type != 'ADAGRAD':
    solver.momentum = 0.9
```

**Comment by lukeyeager** [comment]
Pro tip: after this is ready to merge and when you squash everything into a single commit, you can go ahead and add "Closes #98" to the commit message.

https://help.github.com/articles/closing-issues-via-commit-messages/

Also, please sign the CLA and send it to us in order to get your contribution merged. [Instructions here](https://github.com/NVIDIA/DIGITS/blob/975d25d2ace8c2ad7afb7a5b77d6f37e1d256c72/README.md#bugs-and-feature-requests).

File: digits/model/tasks/caffe_train.py (hunk 1 above)
Comment (lukeyeager) [inline digits/model/tasks/caffe_train.py]: Oh I see what you did. That's clever. Maybe add a comment explaining that you're getting the enum value?

**Comment by drozdvadym** [comment]
Ok, I will fix all your remarks.

And what about the individual CLA (I am  beginner in this question)?

**Comment by lukeyeager** [comment]
As explained [in the README](https://github.com/NVIDIA/DIGITS/blob/975d25d2ace8c2ad7afb7a5b77d6f37e1d256c72/README.md#bugs-and-feature-requests), please print, sign and fill out [this file](https://raw.githubusercontent.com/NVIDIA/DIGITS/master/CLA) and email it to digits [at] nvidia [dot] com.

**Comment by drozdvadym** [comment]
Already done.

**Comment by lukeyeager** [comment]
LGTM.

FYI, I ran some tests with the different solvers to see how they perform:
  1. AlexNet on a subset of the ImageNet dataset (41 categories in 10 groups)
  2. LeNet on MNIST

Solver Type | AlexNet runtime (min) | AlexNet max accuracy | LeNet runtime (min) | LeNet max accuracy 
------------ | ------------- | ------------- | ------------- | -------------
SGD | 115 | 84 | **7.66** | 99.18
AdaGrad | 119 | *50.4* | 7.73 | 99.08
NAG | **114** | **84.4** | 7.72 | **99.21**

That's far from a rigorous test, but it seems like AdaGrad may need some better tuning. I'll merge this for now and look into it later.

**Pull request merged**